### ISSUE TITLE
FreeBSD support

### ISSUE THREAD

**Issue by probonopd** [NatronGitHub/Natron#668]
### Natron version

Any

### Operating system

FreeBSD

### System specs

```markdown
FreeBSD
```

### What were you trying to do?

```markdown
`pkg install natron` on FreeBSD
```

### What did you expect to happen? What happened instead?

I expected it to work. But it did not:

```markdown
FreeBSD% sudo pkg install natron
Passwort: 
Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
All repositories are up to date.
pkg: No packages available to install matching 'natron' have been found in the repositories
```

According to https://www.freshports.org/graphics/natron/, the reason is: "Qt4 has been EOL since december 2015"

### Step-by-step reproduction instructions

On FreeBSD, run:

```
FreeBSD% sudo pkg install natron
```

### Additional details

```markdown
The official Natron download page links to https://www.freshports.org/graphics/natron/, which says that this port is deprecated because "Qt4 has been EOL since december 2015"
```

**Comment by rodlie** [comment]
FreeBSD removed Qt4 and related from ports, nothing we can do about it. You can still build it if you modify the old build scripts (not tested).

Natron should be considered unsupported on FreeBSD, if something don't build we will accept PR's.

**Comment by probonopd** [comment]
Are there plans to retire Qt 4?

**Comment by rodlie** [comment]
Yeah, there are plans to add support for Qt 5.12+, but due to lack of resources it moves down the list of priorities.

I would rather spend the limited time I got on things that improves the application, than spending that time messing with Qt5/Qt6, PySide2+, the SDK and the introductions of new bugs, for minimal improvements (if any).

**Comment by probonopd** [comment]
Understandable. It's a pity that Qt 4 is no longer available in FreeBSD packages.

**Comment by rodlie** [comment]
If we had more users on FreeBSD I could maybe support it (provide binaries), but we never had more than a handful of people interested in using Natron on FreeBSD.

**Comment by probonopd** [comment]
I'd like to contribute my part to change this, by preinstalling Natron in [helloSystem](https://hellosystem.github.io/docs/), a FreeBSD-based desktop system for creators with a focus on simplicity, elegance, and usability. In order to do that, I either need to find a way to get Qt 4 working there again, or wait until Natron is using a more recent Qt version, it seems. Or get the Natron AppImage (for Linux) to run (using the Linuxulator).

**Comment by devernay** [comment]
i guess the build-linux-sdk script could work on freebsd, because there's not much linux-specific in it. it's not just qt4: we also need python2. since we maintain these in the linux sdk, freebsd should share these

we could even easily build it using docker. that way you would have the same versions as on Linux. take a look at tools/docker/natron-sdk/build.sh

what is a good freebsd image to start from? which freebsd version should we target? v11 is eol end of this month

**Comment by rodlie** [comment]
I think only Qt4+PySide is needed, ports still got Python2 (for now), and the depends for the plugins are already there (the plugins are still in ports).

> what is a good freebsd image to start from?

v12 is the current stable.

**Comment by probonopd** [comment]
Since older FreeBSD binaries are _supposed_ to run on newer FreeBSD releases, maybe we can still find Qt 4 packages for an older FreeBSD release [somewhere](http://ftp-archive.freebsd.org/pub/FreeBSD-Archive/old-releases/amd64/9.2-RELEASE/packages/All/). Then one _might_ be able to put together an [application bundle](https://hellosystem.github.io/docs/developer/application-bundles) so as not to "litter" Qt 4 into the installed operating system but privately bundle it with the application. Just a thought, haven't tried it yet.

**Comment by rodlie** [comment]
> Since older FreeBSD binaries are supposed to run on newer FreeBSD releases

Yes, but that excludes everything from ports. An old ports package that depends on other packages from ports will not "just work" on a newer system, as the ports packages has been updated, and probably are not the same anymore.

One option is to resurrect and maintain the old ports build scripts for qt4+pyside and make them work in current ports as an overlay or something.

If we where to start providing binaries for FreeBSD again we will take care of what needs to be included with the app (that's not part of ports).

**Comment by devernay** [comment]
RB-2.5 compiled fine with Qt5 (the Qt5 will will be Natron 2.6).
@probonopd You can make PR's on RB-2.5 to add FreeBSD support.

**Comment by mekanix** [comment]
I'm trying to build Natron based on installation for FreeBSD, but I failed with the following error:
```sh
make[1]: don't know how to make Qt5/NatronEngine/natronengine_module_wrapper.cpp. Stop
```

Any ideas where this file should come from or what did I do wrong?  The commit I'm seeing after clone is 21b2c80a533837602d6a26a6065ebf8f4398d288

**Comment by YakoYakoYokuYoku** [comment]
The last changes in the build system were tested with GNU based Make, not BSD Make. But if you either use GNU Make or give CMake a try (using PR #809) then you might succeed with it.

**Comment by rodlie** [comment]
Does not build on Ubuntu 22.04 either.

``fatal_error: natronengine_python.h: No such file or directory``

**Comment by YakoYakoYokuYoku** [comment]
Then #803 might have break some things up, I'm sorry for that. So this might leave us with CMake as the only option left.

**Comment by rodlie** [comment]
Also tested on Windows/MinGW (latest MSYS2 with Qt/PySide2 5.15.4+kde, gcc 12), and the shiboken stuff fails to generate there as well.

We should probably have some kind of fallback (and set minimum Qt/PySide requirements based on the fallback) since the generation seems to fail on most systems.

**Comment by YakoYakoYokuYoku** [comment]
> Also tested on Windows/MinGW (latest MSYS2 with Qt/PySide2 5.15.4+kde, gcc 12), and the shiboken stuff fails to generate there as well.
>
> We should probably have some kind of fallback (and set minimum Qt/PySide requirements based on the fallback) since the generation seems to fail on most systems.

I'll try to **partially** rollback #803 (because I need the scripts from it), thing is that QMake is actually terrible for this, it's a headache to script and configure. CMake is the **ONLY** way out but PR #809 hasn't been accepted yet. So please in the meantime use that PR when testing Natron with Qt5.

**Comment by rodlie** [comment]
No rush, on my test machine (Ubuntu 22.04) I reverted that commit, so I use the pre-generated files included with Natron.

I will start testing CMake tomorrow.

**Comment by Martinfx** [comment]
natron is back! https://github.com/Martinfx/FreeBSD-Ports/tree/master/natron. I ported Natron to FreeBSD with qt5 but i would like to qt6.

**Comment by mekanix** [comment]
It fails to build on my desktop
```
make
===>   natron-2.5.1.p2 depends on package: boost-libs>0 - found
===>   natron-2.5.1.p2 depends on executable: gsed - found
===>   natron-2.5.1.p2 depends on file: /usr/local/bin/cmake - found
===>   natron-2.5.1.p2 depends on executable: ninja - found
===>   natron-2.5.1.p2 depends on executable: update-desktop-database - found
===>   natron-2.5.1.p2 depends on package: gettext-runtime>=0.22_1 - found
===>   natron-2.5.1.p2 depends on package: pkgconf>=1.3.0_1 - found
===>   natron-2.5.1.p2 depends on file: /usr/local/bin/python3.11 - found
===>   natron-2.5.1.p2 depends on executable: update-mime-database - found
===>   natron-2.5.1.p2 depends on file: /usr/local/bin/ccache - found
===>   natron-2.5.1.p2 depends on shared library: libavformat.so - found (/usr/local/lib/libavformat.so)
===>   natron-2.5.1.p2 depends on shared library: libboost_system.so - found (/usr/local/lib/libboost_system.so)
===>   natron-2.5.1.p2 depends on shared library: libcairo.so - found (/usr/local/lib/libcairo.so)
===>   natron-2.5.1.p2 depends on shared library: libexpat.so - found (/usr/local/lib/libexpat.so)
===>   natron-2.5.1.p2 depends on shared library: libfontconfig.so - found (/usr/local/lib/libfontconfig.so)
===>   natron-2.5.1.p2 depends on shared library: libfreetype.so - found (/usr/local/lib/libfreetype.so)
===>   natron-2.5.1.p2 depends on shared library: libpyside2.cpython-311.so - found (/usr/local/lib/libpyside2.cpython-311.so)
===>   natron-2.5.1.p2 depends on shared library: libshiboken2.cpython-311.so - found (/usr/local/lib/libshiboken2.cpython-311.so)
===>   natron-2.5.1.p2 depends on shared library: libintl.so - found (/usr/local/lib/libintl.so)
===>   natron-2.5.1.p2 depends on shared library: libQt5Concurrent.so - found (/usr/local/lib/qt5/libQt5Concurrent.so)
===>   natron-2.5.1.p2 depends on shared library: libQt5Core.so - found (/usr/local/lib/qt5/libQt5Core.so)
===>   natron-2.5.1.p2 depends on shared library: libQt5Gui.so - found (/usr/local/lib/qt5/libQt5Gui.so)
===>   natron-2.5.1.p2 depends on shared library: libQt5Network.so - found (/usr/local/lib/qt5/libQt5Network.so)
===>   natron-2.5.1.p2 depends on shared library: libQt5OpenGL.so - found (/usr/local/lib/qt5/libQt5OpenGL.so)
===>   natron-2.5.1.p2 depends on shared library: libQt5Widgets.so - found (/usr/local/lib/qt5/libQt5Widgets.so)
===>  Configuring for natron-2.5.1.p2
===>  Performing out-of-source build
/bin/mkdir -p /usr/ports/graphics/natron/work/.build
CMake Warning (dev) at CMakeLists.txt:52 (find_package):
  Policy CMP0167 is not set: The FindBoost module is removed.  Run "cmake
  --help-policy CMP0167" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:2256 (message):
  No header defined for headers; skipping header check (note: header-only
  libraries have no designated component)
Call Stack (most recent call first):
  CMakeLists.txt:52 (find_package)


-- Could NOT find Boost (missing: headers) (found version "1.85.0")
CMake Error at /usr/local/lib/cmake/Qt5Core/Qt5CoreConfig.cmake:14 (message):
  The imported target "Qt5::Core" references the file

     "/usr/local/lib/qt5/bin/qmake"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/local/lib/cmake/Qt5Core/Qt5CoreConfigExtras.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/local/lib/cmake/Qt5Core/Qt5CoreConfigExtras.cmake:9 (_qt5_Core_check_file_exists)
  /usr/local/lib/cmake/Qt5Core/Qt5CoreConfig.cmake:232 (include)
  /usr/local/lib/cmake/Qt5/Qt5Config.cmake:28 (find_package)
  CMakeLists.txt:58 (find_package)


-- Configuring incomplete, errors occurred!
*** Error code 1

Stop.
make: stopped in /usr/ports/graphics/natron
```

Am I doing something wrong?

**Comment by Martinfx** [comment]
No. I am working on fix.

**Comment by Martinfx** [comment]
> It fails to build on my desktop
> 
> ```
> make
> ===>   natron-2.5.1.p2 depends on package: boost-libs>0 - found
> ===>   natron-2.5.1.p2 depends on executable: gsed - found
> ===>   natron-2.5.1.p2 depends on file: /usr/local/bin/cmake - found
> ===>   natron-2.5.1.p2 depends on executable: ninja - found
> ===>   natron-2.5.1.p2 depends on executable: update-desktop-database - found
> ===>   natron-2.5.1.p2 depends on package: gettext-runtime>=0.22_1 - found
> ===>   natron-2.5.1.p2 depends on package: pkgconf>=1.3.0_1 - found
> ===>   natron-2.5.1.p2 depends on file: /usr/local/bin/python3.11 - found
> ===>   natron-2.5.1.p2 depends on executable: update-mime-database - found
> ===>   natron-2.5.1.p2 depends on file: /usr/local/bin/ccache - found
> ===>   natron-2.5.1.p2 depends on shared library: libavformat.so - found (/usr/local/lib/libavformat.so)
> ===>   natron-2.5.1.p2 depends on shared library: libboost_system.so - found (/usr/local/lib/libboost_system.so)
> ===>   natron-2.5.1.p2 depends on shared library: libcairo.so - found (/usr/local/lib/libcairo.so)
> ===>   natron-2.5.1.p2 depends on shared library: libexpat.so - found (/usr/local/lib/libexpat.so)
> ===>   natron-2.5.1.p2 depends on shared library: libfontconfig.so - found (/usr/local/lib/libfontconfig.so)
> ===>   natron-2.5.1.p2 depends on shared library: libfreetype.so - found (/usr/local/lib/libfreetype.so)
> ===>   natron-2.5.1.p2 depends on shared library: libpyside2.cpython-311.so - found (/usr/local/lib/libpyside2.cpython-311.so)
> ===>   natron-2.5.1.p2 depends on shared library: libshiboken2.cpython-311.so - found (/usr/local/lib/libshiboken2.cpython-311.so)
> ===>   natron-2.5.1.p2 depends on shared library: libintl.so - found (/usr/local/lib/libintl.so)
> ===>   natron-2.5.1.p2 depends on shared library: libQt5Concurrent.so - found (/usr/local/lib/qt5/libQt5Concurrent.so)
> ===>   natron-2.5.1.p2 depends on shared library: libQt5Core.so - found (/usr/local/lib/qt5/libQt5Core.so)
> ===>   natron-2.5.1.p2 depends on shared library: libQt5Gui.so - found (/usr/local/lib/qt5/libQt5Gui.so)
> ===>   natron-2.5.1.p2 depends on shared library: libQt5Network.so - found (/usr/local/lib/qt5/libQt5Network.so)
> ===>   natron-2.5.1.p2 depends on shared library: libQt5OpenGL.so - found (/usr/local/lib/qt5/libQt5OpenGL.so)
> ===>   natron-2.5.1.p2 depends on shared library: libQt5Widgets.so - found (/usr/local/lib/qt5/libQt5Widgets.so)
> ===>  Configuring for natron-2.5.1.p2
> ===>  Performing out-of-source build
> /bin/mkdir -p /usr/ports/graphics/natron/work/.build
> CMake Warning (dev) at CMakeLists.txt:52 (find_package):
>   Policy CMP0167 is not set: The FindBoost module is removed.  Run "cmake
>   --help-policy CMP0167" for policy details.  Use the cmake_policy command to
>   set the policy and suppress this warning.
> 
> This warning is for project developers.  Use -Wno-dev to suppress it.
> 
> CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:2256 (message):
>   No header defined for headers; skipping header check (note: header-only
>   libraries have no designated component)
> Call Stack (most recent call first):
>   CMakeLists.txt:52 (find_package)
> 
> 
> -- Could NOT find Boost (missing: headers) (found version "1.85.0")
> CMake Error at /usr/local/lib/cmake/Qt5Core/Qt5CoreConfig.cmake:14 (message):
>   The imported target "Qt5::Core" references the file
> 
>      "/usr/local/lib/qt5/bin/qmake"
> 
>   but this file does not exist.  Possible reasons include:
> 
>   * The file was deleted, renamed, or moved to another location.
> 
>   * An install or uninstall procedure did not complete successfully.
> 
>   * The installation package was faulty and contained
> 
>      "/usr/local/lib/cmake/Qt5Core/Qt5CoreConfigExtras.cmake"
> 
>   but not all the files it references.
> 
> Call Stack (most recent call first):
>   /usr/local/lib/cmake/Qt5Core/Qt5CoreConfigExtras.cmake:9 (_qt5_Core_check_file_exists)
>   /usr/local/lib/cmake/Qt5Core/Qt5CoreConfig.cmake:232 (include)
>   /usr/local/lib/cmake/Qt5/Qt5Config.cmake:28 (find_package)
>   CMakeLists.txt:58 (find_package)
> 
> 
> -- Configuring incomplete, errors occurred!
> *** Error code 1
> 
> Stop.
> make: stopped in /usr/ports/graphics/natron
> ```
> 
> Am I doing something wrong?

@mekanix You try now https://github.com/Martinfx/FreeBSD-Ports/tree/master/natron

**Comment by mekanix** [comment]
It builds, but I had to remove `share/pixmaps/natronProjectIcon_linux.png` from pkg-plist.
